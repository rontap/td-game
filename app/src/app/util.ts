type jsobj = {
    [key: string]: any
}

interface Ijsobj {
    [key: string]: any
}

export type {
    jsobj,
    Ijsobj
}


export const preventBubble = (fn: Function) => (evt: any) => {
    evt.preventDefault();
    evt.stopPropagation();
    fn(evt);
}
// // @ts-ignore
// Map.prototype.map = function (fn: function, i: ?number) {
//     console.log('??');
//     return [...(this.values())].map(fn, i);
// }
// console.log('?34');
// const mcguffin = {};
// export {mcguffin};