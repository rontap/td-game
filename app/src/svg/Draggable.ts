enum Button {
    DOWN = "DOWN",
    UP = "UP",
    MOVE = "MOVE",
    LEAVE = "LEAVE"
}

class Point {
    public x: number;
    public y: number;

    constructor(x: number, y: number) {
        this.x = x;
        this.y = y;
    }

    unwrap() {
        return {
            x: this.x,
            y: this.y
        }
    }


    static get Origin() {
        return new Point(0, 0);
    }

    add(x: number = 0, y: number = 0) {
        return new Point(this.x + x, this.y + y);
    }

}

class Geom {
    static Distance(a: Point, b: Point) {

    }

    static Difference(a: Point, b: Point): Point {
        return new Point(
            a.x - b.x,
            a.y - b.y
        )
    }

    static Inv(a: Point) {
        return new Point(
            -a.x,
            -a.y
        )
    }
}

const moveableElements = ["svg", "rect", "foreignObject"];

class DragHandler {
    private isDown: boolean = false;
    private ctmx: SVGMatrix | undefined;

    constructor() {
    }

    public startCoord: Point = Point.Origin;
    public currentCoord: Point = Point.Origin;
    public selected: EventTarget | null = null;

    bubbleEvt(target: any): any {

        if (moveableElements.includes(target.tagName)) {
            return target;
        } else {
            return this.bubbleEvt(target.parentElement);
        }

    }

    evt(action: string, evt: any) {
        this.getTransformMatrix();
        evt.preventDefault();
        evt.target = this.bubbleEvt(evt.target);

        const id: number = Number(evt.target.getAttribute('data-id'));

        if (evt.target.tagName !== 'svg') {
            evt.stopPropagation();
        }

        if (action === Button.DOWN) {
            this.selected = evt.target;
            this.startCoord = this.getCursor(evt);
            this.startCoord = Geom.Difference(this.startCoord, DragHandler.getCoords(evt.target));
            this.isDown = true;
        }

        if (action === Button.MOVE) {
            if (this.selected && this.isDown && this.ctmx) {
                const finalCoord = Geom.Difference(this.getCursor(evt), this.startCoord);
                this.setCoords(
                    this.selected,
                    finalCoord
                );

                // const line: Element = document.querySelector(".data-line-0-1")!;
                //
                // this.setCoords(line, finalCoord, 'x1', 'y1');
                // this.setCoords(line, DragHandler.getCoords(
                //     document.querySelector(".f2")!
                // ), 'x2', 'y2');

                document.querySelectorAll('.data-node-from-' + id).forEach(item => {
                    this.setCoords(item, finalCoord.add(50, 40), 'x1', 'y1');
                })

                document.querySelectorAll('.data-node-to-' + id).forEach(item => {
                    this.setCoords(item, finalCoord.add(0,40), 'x2', 'y2');
                })

            }


        }


        if (action === Button.UP || action === Button.LEAVE) {
            this.isDown = false;
        }

    }

    getTransformMatrix() {
        //@ts-ignore
        this.ctmx = document.querySelector(".svgRoot").getScreenCTM();
    }

    get SvgElement() {
        return document.querySelector(".svgRoot");
    }

    getCursor(evt: any) {
        if (!this.ctmx) throw Error('no transform mx');

        return new Point(
            (evt.clientX - this.ctmx.e) / this.ctmx.a,
            (evt.clientY - this.ctmx.f) / this.ctmx.d)
    }

    static getCoords(item: any) {
        if (!item) {
            return Point.Origin;
        }
        return new Point(
            Number(item.getAttributeNS(null, 'x')),
            Number(item.getAttributeNS(null, 'y'))
        )
    }

    setCoords(item: any, point: Point, xName = 'x', yName = 'y') {
        item.setAttributeNS(null, xName, point.x + EPS);
        item.setAttributeNS(null, yName, point.y + EPS);
    }
}

const Draggable = {
    onMouseDown: (evt: any) => DragHandlerInst.evt(Button.DOWN, evt),
    onMouseUp: (evt: any) => DragHandlerInst.evt(Button.UP, evt),
    onMouseMove: (evt: any) => DragHandlerInst.evt(Button.MOVE, evt),
    onMouseLeave: (evt: any) => DragHandlerInst.evt(Button.LEAVE, evt),
};


const DragHandlerInst = new DragHandler();
const EPS = 0.00001;
// @ts-ignore
window.dh = DragHandlerInst;

export default Draggable;
export {DragHandlerInst, DragHandler};