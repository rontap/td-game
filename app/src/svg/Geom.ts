
export enum Button {
    DOWN = "DOWN",
    UP = "UP",
    MOVE = "MOVE",
    LEAVE = "LEAVE"
}

export class Point {
    public x: number;
    public y: number;

    constructor(x: number, y: number) {
        this.x = x;
        this.y = y;
    }

    unwrap() {
        return {
            x: this.x,
            y: this.y
        }
    }

    static get Origin() {
        return new Point(0, 0);
    }

    add(x: number = 0, y: number = 0) {
        return new Point(this.x + x, this.y + y);
    }

    subtract(x: number = 0, y: number = 0) {
        return new Point(this.x - x, this.y - y);
    }

}

export class Geom {
    static Distance(a: Point, b: Point) {

    }

    static Difference(a: Point, b: Point): Point {
        return new Point(
            a.x - b.x,
            a.y - b.y
        )
    }

    static Inv(a: Point) {
        return new Point(
            -a.x,
            -a.y
        )
    }

    static bezierSvgD(fromPoint: Point, toPoint: Point): string {
        const sx = fromPoint.x;
        const ex = toPoint.x;
        const mx = (sx + ex) / 2;

        const sy = fromPoint.y;
        const ey = toPoint.y;
        const my = (sy + ey) / 2;

        return `M${sx},${sy} C${mx},${sy} ${mx},${ey} ${ex},${ey}`
    }
}