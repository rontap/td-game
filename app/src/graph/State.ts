import {jsobj} from "../app/util";
import {create, useStore} from 'zustand'
import {devtools, persist} from 'zustand/middleware'
import {Node} from "../node/Node";
import {Line, LineId, NodeId} from "../node/Line";
import {unwatchFile} from "fs";
import {temporal, TemporalState} from 'zundo'
import {shallow} from "zustand/shallow";
import {GraphUtil, GraphUtilInst} from "./GraphUtil";

// export default class State {
//     static nodes: Node[] = [];
// }

interface AppState {
    nodes: Node[],
    lines: Line[],
    addNode: (node: Node) => void
    getNodeById: (id: number) => Node | undefined
    removeNode: (id: number) => void,
    addLine: (line: Line) => boolean,
    removeLine: (id: number) => void,

    resetStore: () => void,

    blueprintedNode: string,
    getLineBetween: (from: number, to: number) => Line | undefined
    zoom: number,
    contextMenu: any,
    lineAddAt: Node | undefined,
    activeNode: Node | undefined,
    setActiveNode: (id?: NodeId | undefined) => void,
    getLinesAtNodeConnection: (id: NodeId | undefined, end: End) => Line[],
    setBlueprintedNode: (nodeName: string) => void,
    forceSvgRender: {},
    doSvgRender: () => void,

    setSingleEdgeOfActiveNode:
        (propertyName: string, to: string) => void,
}

export enum End { FROM, TO}

const initialState = {
    lines: [],
    lineAddAt: undefined,
    contextMenu: {},
    blueprintedNode: "",
    activeNode: undefined,
    forceSvgRender: {},
    nodes: [],
    zoom: 1,
}

const actions = {}

const State = create<AppState>()(
    devtools(
        temporal(
            //persist(
            (set, get) => ({
                setSingleEdgeOfActiveNode: (propertyName: string, to: string) =>
                    set((state) => {
                            const activeNode = state.activeNode;
                            console.log('->>', activeNode, to);
                            if (!activeNode) return {};
                            activeNode._configurableInputValues.set(propertyName, to);
                            return {activeNode}
                        }
                    ),
                addNode: (node: Node) => set((state) =>
                    ({nodes: state.nodes.concat(node)})),
                getNodeById: (id: NodeId) =>
                    get().nodes.find(item => item.ID === Number(id)),
                removeNode: (id: NodeId) => set((state) =>
                    ({nodes: state.nodes.filter(item => item.ID !== Number(id))})),
                addLine: (line: Line) => {
                    if (get().getLineBetween(line.from, line.to)) {
                        return false;
                    }
                    set((state) => ({lines: state.lines.concat(line)}));
                    GraphUtilInst.detectCircles();
                    return true;
                },
                removeLine: (id: LineId) => {
                    set((state) => ({lines: state.lines.filter(item => item.ID !== Number(id))}));
                    GraphUtilInst.detectCircles();
                },
                getLineBetween: (from: number, to: number) =>
                    get().lines.find(line => line.from === from && line.to === to),
                setActiveNode: (id) => set((state) =>
                    ({activeNode: id ? get().nodes.find(item => item.ID === Number(id)) : undefined})
                ),
                setBlueprintedNode: (nodeName: string) => set((state) =>
                    ({blueprintedNode: nodeName})),
                getLinesAtNodeConnection: (id: NodeId | undefined, end: End) => {
                    if (!id) return [];
                    const whichJunction = end === End.FROM ? "from" : "to";
                    return get().lines.filter(line => line[whichJunction] === id)
                },
                doSvgRender: () => set((state) =>
                    ({forceSvgRender: {}})
                ),
                resetStore: () => set((state) => initialState),

                // store part
                ...initialState
            }),
            {
                partialize: (state) => {
                    const {lines, nodes, ...rest} = state
                    return {lines, nodes}
                },
                equality: shallow
            }
            //)
        ) // temporal
        , {
            name: 'store',
        }
    )
);

const useTemporalStore = <T, >(
    selector: (state: TemporalState<any>) => T,
    equality?: (a: T, b: T) => boolean,
) => useStore(State.temporal, selector, equality);

export default State;
const {getState, setState, subscribe} = State
export {
    getState,
    setState,
    subscribe,
    useTemporalStore
};
// @ts-ignore
window.State = State;