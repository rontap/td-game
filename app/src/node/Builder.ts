import {loadJsonNodeDefinitions, NodeTemplate, NodeTemplateMap} from "../app/DynamicReader";
import {EdgeInvariant, EdgeLoader, edgeTypes} from "../graph/EdgeLoader";
import {jsobj} from "../app/util";
import {Node} from './Node';
import State from "../graph/State";

EdgeLoader();

export class NodeBuilder {
    static get rawTypes(): NodeTemplateMap {
        return this._rawTypes;
    }

    static get types(): NodeTemplateMap {
        return this._types;
    }

    private static _rawTypes = new Map<string, NodeTemplate>();
    private static _types = new Map<string, NodeTemplate>();

    static Build() {
        this._rawTypes = loadJsonNodeDefinitions();

        [...this._rawTypes.values()].map((value: NodeTemplate) => {
            this._types.set(value.name, value);
        })

        return this._rawTypes;
    }

    static New(nodeType: string) {
        const nodes = State.getState().nodes;
        State.setState({nodes: nodes.concat(new Node(nodeType))})
    }

    static getType(name: string) {
        return this.types.get(name);
    }

    // static InstantiateNode(nodeTemplate: jsobj) {
    //     switch (nodeTemplate.type) {
    //         case "default":
    //             return new Node(nodeTemplate, nodeTemplate.type);
    //         case undefined:
    //             throw Error("Node type property is required for instantiation.");
    //         default:
    //             throw Error("Unknown Node type: " + nodeTemplate.type);
    //     }
    // }

}

EdgeInvariant(true);