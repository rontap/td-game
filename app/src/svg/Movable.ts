import {jsobj} from "../util/util";
import CONST from "../const";
import State from "../graph/State";
import {Line} from "../node/Line";

import {Node} from "../node/Node";
import {DragHandler, DragHandlerInst} from "./Draggable";
import {Geom, Point} from '../util/Geom';

function initMovable() {

}

let svgImageAct = document.querySelector(".svgRoot")!;
//const svgContainer = document.getElementById("svgContainer");
const svgImage = CONST.rectSize;
let viewBox = {
    x: 0,
    y: 0,
    w: svgImage.clientWidth,
    h: svgImage.clientHeight
};
svgImageAct?.setAttribute('viewBox', Geom.viewBox(viewBox));
const svgSize = {
    w: svgImage.clientWidth,
    h: svgImage.clientHeight
};
let startPoint = Point.Origin;
let endPoint = Point.Origin;
let scale = 1;

class MovableStateClass {
    constructor() {

    }

    get zoomLevel(): number {
        return State.getState().zoom;
    }

    set zoomLevel(zoom: number) {
        State.setState({zoom});
    }

    public lineAdd: Node | undefined = undefined;

    public isPanning: boolean = false;

    resetZoom() {
        viewBox = {x: 0, y: 0, w: svgImage.clientWidth, h: svgImage.clientHeight};
        svgImageAct?.setAttribute('viewBox', Geom.viewBox(viewBox));
        MovableState.zoomLevel = 1;
    }

    zoom(mx: number, my: number, direction: number) {

        // prevent uner and over zooming.
        if ((MovableState.zoomLevel < CONST.zoom.min && direction === -1)
            || (MovableState.zoomLevel > CONST.zoom.max && direction === 1)
        ) return false;

        svgImageAct = document.querySelector(".svgRoot")!;

        // svg viewbox visszaállítása az alapértelmezetre
        svgImageAct?.setAttribute('viewBox', Geom.viewBox(viewBox));
        // dw és dh zoom méretét szabják meg. A direction az -1 vagy +1
        let dw = viewBox.w * direction * CONST.zoom.speed;
        let dh = viewBox.h * direction * CONST.zoom.speed;
        // mx és my értékek a kurzor jelenelegi pozíciója. Így érhető el az, hogy a zoomolás az a kurzor felé történjen
        let dx = dw * mx / svgSize.w;
        let dy = dh * my / svgSize.h;
        // az új viewbox kiszámolása az alapértelmezett viewboxtól
        viewBox = {
            x: viewBox.x + dx,
            y: viewBox.y + dy,
            w: viewBox.w - dw,
            h: viewBox.h - dh
        };
        scale = svgSize.w / viewBox.w;
        MovableState.zoomLevel = scale;
        svgImageAct?.setAttribute('viewBox', Geom.viewBox(viewBox));
    }

    zoomCenter(direction: any) {
        svgImageAct = document.querySelector(".svgRoot")!;
        const {width, height} = svgImageAct.getBoundingClientRect();
        this.zoom(width / 2, height / 2, direction);
    }

    beginLineAdd(node: Node) {
        if (this.lineAdd == undefined) {
            this.lineAdd = node;
            State.setState({lineAddAt: node})
            return;
        }
    }

    finishLineAdd(node: Node) {
        if (this.lineAdd != undefined) {
            Line.New(this.lineAdd.ID, node.ID);
        }

        this.endLineAdd();
    }

    endLineAdd() {
        State.setState({lineAddAt: undefined});
        this.lineAdd = undefined;
    }

}

// @ts-ignore
window._SVGViewBox = viewBox;
export const MovableState = new MovableStateClass();
const svgContainer: jsobj = {};

svgContainer.onWheel = function (e: any) {
    MovableState.zoom(e.nativeEvent.offsetX, e.nativeEvent.offsetY, Math.sign(-e.deltaY))
}

svgContainer.onContextMenu = function (e: any) {
    e.preventDefault();
    if (MovableState.lineAdd) {
        MovableState.endLineAdd();
        return;
    }
    State.setState({contextMenu: e});
}


svgContainer.onMouseDown = function ({nativeEvent: e}: any) {
    MovableState.isPanning = true;
    startPoint = new Point(e.x, e.y);
}

svgContainer.onMouseMove = function ({nativeEvent: e}: any) {
    if (MovableState.isPanning) {
        State.setState({contextMenu: e});
        svgImageAct = document.querySelector(".svgRoot")!;
        endPoint = new Point(e.x, e.y);
        let dx = (startPoint.x - endPoint.x) / scale;
        let dy = (startPoint.y - endPoint.y) / scale;
        let movedViewBox = {x: viewBox.x + dx, y: viewBox.y + dy, w: viewBox.w, h: viewBox.h};
        svgImageAct?.setAttribute('viewBox', Geom.viewBox(movedViewBox));
    }

    if (MovableState.lineAdd) {
        setLineAddCoords(MovableState.lineAdd, e);
    }
}

svgContainer.onMouseUp = function ({nativeEvent: e}: any) {
    if (MovableState.isPanning) {
        svgImageAct = document.querySelector(".svgRoot")!;
        endPoint = new Point(e.x, e.y);
        let dx = (startPoint.x - endPoint.x) / scale;
        let dy = (startPoint.y - endPoint.y) / scale;
        viewBox = {x: viewBox.x + dx, y: viewBox.y + dy, w: viewBox.w, h: viewBox.h};
        svgImageAct?.setAttribute('viewBox', Geom.viewBox(viewBox));
        MovableState.isPanning = false;
    }
}

svgContainer.onClick = function (e: any) {
    State.setState({contextMenu: e});
}

svgContainer.onMouseEnter = function (e: any) {
    MovableState.isPanning = false;
}

svgContainer.onMouseLeave = function (e: any) {
    MovableState.isPanning = false;
}

const setLineAddCoords = (fromNode: Node, toEvt: Event) => {
    const fromPoint = DragHandler
        .getCoords(fromNode.selfSvg)
        .add(CONST.box.width + CONST.box.padLeft, CONST.box.pointTop);
    const toPoint = DragHandlerInst
        .getCursor(toEvt)
        .add(CONST.box.padLeft, 0);

    document.querySelector('.currentBez')
        ?.setAttributeNS(
            null,
            'd',
            Geom.bezierSvgD(fromPoint, toPoint)
        )
}

export default svgContainer;