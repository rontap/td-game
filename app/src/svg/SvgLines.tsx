import {Line} from "../node/Line";
import State from "../graph/State";

export default function SvgLines(props: any) {
    const lines = State((state) => state.lines)
    const tempSvgRender = State((state) => state.forceSvgRender)

    return <>
        {lines.map((node: Line) => node.getSvg())}
    </>
}