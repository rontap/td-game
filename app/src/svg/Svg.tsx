import CONST from "../const";
import Draggable, {DragHandler, DragHandlerInst, Geom, Point} from "./Draggable";
import {useEffect} from "react";
import {jsobj} from "../app/util";
import State, {getState} from "../graph/State";

import {Node} from "../node/Node";
import Movable from "./Movable.js";
import {Line} from "../node/Line";
import SvgLines from "./SvgLines";

export default function Svg(props: jsobj) {
    const {blueprint} = props;
    useEffect(() => {
        DragHandlerInst.getTransformMatrix();
    }, []);

    let nodes, lines, tempSvgRender;


    console.log('rendering svg')
    nodes = State((state) => state.nodes)


    return (
        <div id={"svgRootCont"}
             {...Movable}
        >
            <svg
                {...Draggable}
                {...(blueprint ? CONST.blueprintRectSize : CONST.rectSize)}
                className={"svgRoot"}
            >

                <defs>
                    <pattern id="smallGrid" width="8" height="8" patternUnits="userSpaceOnUse">
                        <path d="M 8 0 L 0 0 0 8" fill="none" stroke="#384850" strokeWidth="0.5"/>
                    </pattern>
                    <pattern id="grid" width="80" height="80" patternUnits="userSpaceOnUse">
                        <rect width="80" height="80" fill="url(#smallGrid)"/>
                        <path d="M 80 0 L 0 0 0 80" fill="none" stroke="#5b6d76" strokeWidth="1"/>
                    </pattern>
                </defs>


                <defs>
                    <marker
                        id="dot"
                        viewBox="0 0 10 10"
                        refX="5"
                        refY="5"
                        markerWidth="5"
                        markerHeight="5">
                        <circle cx="5" cy="20" r="20" fill="red"/>
                        {/*<animateMotion*/}

                        {/*    dur="3s"*/}

                        {/*    repeatCount="indefinite"*/}
                        {/*    path={Geom.bezierSvgD(fromPoint, toPoint)}/>*/}

                    </marker>
                </defs>

                <rect
                    x="-100%"
                    y="-100%"
                    width="1000%"
                    height="10000%" fill="url(#grid)"/>

                {/*<rect className="draggable f2" x="160" y="190" width="80" height="40"*/}

                {/*      fill="#5577ff">*/}
                {/*</rect>*/}

                {/*<rect className="draggable" x="350" y="210" width="80" height="40"*/}

                {/*      fill="#ff77aa">*/}
                {/*</rect>*/}


                {
                    getCurrentLine()
                }

                <SvgLines/>


                {
                    nodes.map((node: Node) => node.getSvg())
                }

                {/*{*/}
                {/*    nodes.map((node: Node) => node.getInputLines())*/}
                {/*}*/}


                {/*<path d="M100,100 C250,100 250,250 400,250"*/}
                {/*      style={{fill: 'transparent', stroke: 'red'}}/>*/}

                {/*<line*/}
                {/*    className={"data-line-0-1"}*/}
                {/*    x1="0" y1="80" x2="100" y2="20" stroke="black"/>*/}

                {/*<foreignObject x="200" y="20" width="160" height="100">*/}
                {/*    <div className={"boxedItem"}>*/}
                {/*        hi<br/>*/}
                {/*        <button>btn</button>*/}
                {/*    </div>*/}
                {/*</foreignObject>*/}


                {/*<defs>*/}

                {/*    <marker*/}
                {/*        id="dot"*/}
                {/*        viewBox="0 0 10 10"*/}
                {/*        refX="5"*/}
                {/*        refY="5"*/}
                {/*        markerWidth="5"*/}
                {/*        markerHeight="5">*/}
                {/*        <circle cx="5" cy="5" r="5" fill="red"/>*/}
                {/*    </marker>*/}
                {/*</defs>*/}

                {/*<polyline*/}
                {/*    points="15,80 29,50 43,60 57,30 71,40 85,15"*/}
                {/*    fill="none"*/}
                {/*    stroke="grey"*/}
                {/*    markerStart="url(#dot)"*/}
                {/*    markerMid="url(#dot)"*/}
                {/*    markerEnd="url(#dot)"/>*/}

                {/*<path*/}
                {/*    d="M30.367029359882025,222.74918426476106 C132.1095737779142,222.74918426476106 132.1095737779142,104.44504125761989 233.85211819594636,104.44504125761989"*/}
                {/*    className="data-curve-from-2 data-curve-to-1" x1="30.367029359882025" y1="222.74918426476106"*/}
                {/*    x2="233.85212819594636" y2="104.4450512576199"*/}
                {/*    markerStart="url(#dot)"*/}
                {/*    markerMid="url(#dot)"*/}
                {/*    markerEnd="url(#dot)"*/}
                {/*    style={{fill: "transparent", stroke: "whitesmoke", width: "3px"}}*/}
                {/*    path="M30.367029359882025,222.74918426476106 C132.1095737779142,222.74918426476106 132.1095737779142,104.44504125761989 233.85211819594636,104.44504125761989"></path>*/}

            </svg>
        </div>
    );
}
const getCurrentLine = () => {
    const lineAddAt = State((state) => state.lineAddAt)
    const node = State((state) => state.getNodeById(lineAddAt?.id || -1))

    if (!node?.coords) {
        return <></>
    }

    if (!lineAddAt.id) {
        return <></>
    }

    const fromPoint = DragHandler.getCoords(node.selfSvg).add(CONST.box.width, CONST.box.pointTop);
    const toPoint = DragHandlerInst.getCursor(lineAddAt.evt);

    return <>
        <path d={Geom.bezierSvgD(fromPoint, toPoint)}
              className={"currentBez currentNodeActive"}/>
    </>
}