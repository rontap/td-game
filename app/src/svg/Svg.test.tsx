import CONST from "../const";
import Draggable, {DragHandler, DragHandlerInst} from "./Draggable";
import {useEffect} from "react";
import {jsobj} from "../app/util";
import State, {getState} from "../graph/State";
import {Geom, Point} from "../geometry/Geom";
import {Node} from "../node/Node";
import Movable from "./Movable.js";
import {Line} from "../node/Line";
import SvgLines from "./SvgLines";
import Svg from "./Svg";
import {NodeBuilder} from "../node/Builder";
// @ts-ignore
import renderer from 'react-test-renderer';

it('Renders <Svg/> without issues', () => {
    const svgCore = renderer.create(<Svg items={NodeBuilder.Build()}/>)
        .toJSON();
    expect(svgCore).toMatchSnapshot();
})