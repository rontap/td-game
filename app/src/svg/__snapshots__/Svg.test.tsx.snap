// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders <Svg/> without issues 1`] = `
<div
  id="svgRootCont"
  onClick={[Function]}
  onContextMenu={[Function]}
  onMouseDown={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  onMouseMove={[Function]}
  onMouseUp={[Function]}
  onWheel={[Function]}
>
  <svg
    className="svgRoot"
    clientHeight={768}
    clientWidth={1024}
    height={768}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseMove={[Function]}
    onMouseUp={[Function]}
    width={1024}
    x={0}
    xmlns="http://www.w3.org/2000/svg"
    y={0}
  >
    <defs>
      <pattern
        height="8"
        id="smallGrid"
        patternUnits="userSpaceOnUse"
        width="8"
      >
        <path
          d="M 8 0 L 0 0 0 8"
          fill="none"
          stroke="#384850"
          strokeWidth="0.5"
        />
      </pattern>
      <pattern
        height="80"
        id="grid"
        patternUnits="userSpaceOnUse"
        width="80"
      >
        <rect
          fill="url(#smallGrid)"
          height="80"
          width="80"
        />
        <path
          d="M 80 0 L 0 0 0 80"
          fill="none"
          stroke="#5b6d76"
          strokeWidth="1"
        />
      </pattern>
    </defs>
    <defs>
      <marker
        id="dot"
        markerHeight="5"
        markerWidth="5"
        refX="5"
        refY="5"
        viewBox="0 0 10 10"
      >
        <circle
          cx="5"
          cy="20"
          fill="red"
          r="20"
        />
      </marker>
    </defs>
    <rect
      fill="url(#grid)"
      height="10000%"
      id="bgRectSvg"
      width="1000%"
      x="-100%"
      y="-100%"
    />
  </svg>
</div>
`;
